scripts/irods/test/test_dynamic_peps.py

```diff
@@ -199,3 +201,107 @@ def test_data_obj_info_parameters_in_pep_database_reg_data_obj_post__issue_5554(
                     self.admin.assert_icommand(['irm', '-f', logical_path])
                     self.admin.assert_icommand(['iadmin', 'rum'])
 
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-irods_rule_language', 'Only implemented for NREP.')
+    @unittest.skipIf(test.settings.TOPOLOGY_FROM_RESOURCE_SERVER , 'Database PEPs only fire on the catalog provider.')
+    def test_pep_database_data_object_finalize__issue_6385(self):
+        other_resource = 'test_pep_database_data_object_finalize_resc'
+        filename = 'test_pep_database_data_object_finalize__issue_6385'
+        contents = 'knock knock'
+        logical_path = os.path.join(self.user.session_collection, filename)
+        expected_attribute = 'latest_replica'
+
+        pep_map = {
+            'irods_rule_engine_plugin-irods_rule_language': dedent('''
+                pep_database_data_object_finalize_post(*instance, *ctx, *out, *replicas) {{
+                    # This PEP annotates an AVU to the target data object indicating the most recently created replica.
+
+                    # Get a JSON handle so we can work with the replicas input.
+                    msiStrlen(*replicas, *size);
+                    msi_json_parse(*replicas, int(*size), *handle);
+
+                    # Get the logical path based on the data ID. (Seems silly that we have to do this.)
```

Comment:
Any ideas on how to improve it?

Feel free to create an issue for it.

Reply:
We could change the database operation to take a different JSON structure...

Right now, it just takes an array of JSON objects representing the "before" and "after" states of R_DATA_MAIN rows. However, the data object finalize API has a JSON structure which holds a few more inputs, including the logical path, which it could just pass along to this as part of the input. 

Perhaps we can make the `_replicas` input a more general-purpose input:
```js
{
    "logical_path": "/tempZone/home/rods/foo",
    "replicas": [
        {
            "before" : {
                "data_id": "10018",
                ...
            },
            "after": {
                "data_id": "10018",
                ...
            }
        },
        ...
    ]
}
```

Reply:
My first thought is, does that play well with the goals of the finalize API?

I'm open to discussing this more. Capture your ideas in an issue so we don't forget.

Reply:
I think we should at least make this a JSON object with a way to key into the replicas array rather than just passing an array. Changing the interface (that is, the JSON structure) later will get us into a similar situation as the one described in #6385 to begin with (i.e. affecting how administrators implement policy). If we go ahead and make it a JSON object with one member, the logical path component can be added later without affecting existing policy implementations. Thoughts?

I'll create an issue once we've reached a consensus on the interface vis a vis what needs to be done before the release.

Reply:
I see. That sounds good. Please proceed.

Reply:
I created https://github.com/irods/irods/issues/7031 for adding the logical path component to the input. I changed the input to be a JSON object with a single key, "replicas", which points to the array. I also expanded the Doxygen to describe it in more detail rather than referring to the other file.

Since adding the logical path component at this time would be non-trivial, I think we should punt on that part.

Reply:
I agree.